finetrainers/trainer.py

```diff
@@ -111,36 +176,189 @@ def prepare_models(self) -> None:
+
+        if self.args.caption_dropout_p > 0 and self.args.caption_dropout_technique == "empty":
+            logger.warning(
+                "Caption dropout is not supported with precomputation yet. This will be supported in the future."
+            )
+
+        conditions_dir = Path(self.args.data_root) / PRECOMPUTED_DIR_NAME / PRECOMPUTED_CONDITIONS_DIR_NAME
+        latents_dir = Path(self.args.data_root) / PRECOMPUTED_DIR_NAME / PRECOMPUTED_LATENTS_DIR_NAME
+        conditions_dir.mkdir(parents=True, exist_ok=True)
+        latents_dir.mkdir(parents=True, exist_ok=True)
+
+        # Precompute conditions
+        progress_bar = tqdm(
+            range(0, len(self.dataset)),
+            desc="Precomputing conditions",
+            disable=not self.state.accelerator.is_local_main_process,
+        )
+        index = 0
+        for i, data in enumerate(self.dataset):
+            if i % self.state.accelerator.num_processes != self.state.accelerator.process_index:
+                continue
+
+            logger.debug(
+                f"Precomputing conditions and latents for batch {i + 1}/{len(self.dataset)} on process {self.state.accelerator.process_index}"
+            )
+
+            other_conditions = self.model_config["prepare_conditions"](
+                tokenizer=self.tokenizer,
+                tokenizer_2=self.tokenizer_2,
+                tokenizer_3=self.tokenizer_3,
+                text_encoder=self.text_encoder,
+                text_encoder_2=self.text_encoder_2,
+                text_encoder_3=self.text_encoder_3,
+                prompt=data["prompt"],
+                device=self.state.accelerator.device,
+                dtype=self.state.weight_dtype,
+            )
+            filename = conditions_dir / f"conditions-{i}-{index}.pt"
+            torch.save(other_conditions, filename.as_posix())
+            index += 1
+            progress_bar.update(1)
+        self._nuke_components()
+
+        memory_statistics = get_memory_statistics()
+        logger.info(f"Memory after precomputing conditions: {json.dumps(memory_statistics, indent=4)}")
+        torch.cuda.reset_peak_memory_stats(self.state.accelerator.device)
+
+        # Precompute latents
+        latent_components = self.model_config["load_latent_models"](**self._get_load_components_kwargs())
+        self._set_components(latent_components)
+        self._move_components_to_device()
+
+        # TODO(aryan): refactor later
+        components_to_disable_grads = [self.vae]
+        for component in components_to_disable_grads:
+            if component is not None:
+                component.requires_grad_(False)
+
+        if self.vae is not None:
+            if self.args.enable_slicing:
+                self.vae.enable_slicing()
+            if self.args.enable_tiling:
+                self.vae.enable_tiling()
+
+        progress_bar = tqdm(
+            range(0, len(self.dataset)),
+            desc="Precomputing latents",
```

Comment:
We can look into condition precomputation part and design for it later. What we should aim for is:
- Text encoding related parts don't load VAE
- VAE related parts don't load text encoder

So what we now call "precomputing latents" is basically "prepare all things related to VAE", and "precomputing conditions" is basically "prepare all things related to text encoders". Additionally, we might have other things like control condition preparations, which would go other VAE-related stuff. Any other kind of embeddings, like facial image embeds, can go into a separate dedicated method that does so in a way that doesn't simultaneously load text encoder or vae at the same time. We should also ensure that the condition preparation methods should be independent of accelerator states, so that in the future we can implement FSDP and tensor parallelism without having to refactor.

>  Later, we could also make it so that each of the above is executed on different GPUs should a multi-GPU configuration is discovered (and based on user preferences).

Currently, we first precompute all embeddings distributed across GPUs first, and then precompute all latents across all GPUs. Is this not how we want it?

Reply:
Okay. Computing captions are independent from computing latents. So, this can happen parallelly. But no biggie.